File: ios/Classes/Interop/Action.swift
enum Action: String {
    case optIn = "optIn"
    case optOut = "optOut"
    case optOutAndWipeData = "optOutAndWipeData"
    
    case push_getLastKnownPushToken = "push.getLastKnownPushToken"
    case push_RequestPermission = "push.requestPermission"
    case push_iOSRequestProvisionalPermission = "push.iOS.requestProvisionalPermission"
    case push_iOSRefreshToken = "push.iOS.refreshToken"
    case push_clearBadge = "push.clearBadge"
    case push_dismissNotifications = "push.dismissNotifications"
    case push_iOSSetShowForegroundNotifications = "push.iOS.setShowForegroundNotifications"
    
    case user_getIdentifier = "user.getIdentifier"
    case user_getLanguage = "user.getLanguage"
    case user_getRegion = "user.getRegion"
    case user_getInstallationID = "user.getInstallationID"
    case user_fetchAttributes = "user.fetch.attributes"
    case user_fetchTags = "user.fetch.tags"
    case user_edit = "user.edit"
    case user_trackEvent = "user.track.event"
    case user_trackTransaction = "user.track.transaction"
    case user_trackLocation = "user.track.location"
    
    case inbox_createInstallationFetcher = "inbox.createInstallationFetcher"
    case inbox_createUserFetcher = "inbox.createUserFetcher"
    case inbox_releaseFetcher = "inbox.releaseFetcher"
    case inbox_fetchNewNotifications = "inbox.fetchNewNotifications"
    case inbox_fetchNextPage = "inbox.fetchNextPage"
    case inbox_getFetchedNotifications = "inbox.getFetchedNotifications"
    case inbox_markAsRead = "inbox.markAsRead"
    case inbox_markAllAsRead = "inbox.markAllAsRead"
    case inbox_markAsDeleted = "inbox.markAsDeleted"
    case inbox_displayLandingMessage = "inbox.displayLandingMessage"

    case messaging_setDoNotDisturbEnabled = "messaging.setDoNotDisturbEnabled"
    case messaging_showPendingMessage = "messaging.showPendingMessage"
    
    case debug_showDebugView = "debug.showDebugView"
    
    case echo = "echo"
}

File: android/src/main/java/com/batch/batch_flutter/interop/Action.java
package com.batch.batch_flutter.interop;

/**
 * Enum that wraps the supported action names.
 */
enum Action
{
    OPT_IN("optIn"),
    OPT_OUT("optOut"),
    OPT_OUT_AND_WIPE_DATA("optOutAndWipeData"),
    MESSAGING_SET_DO_NOT_DISTURB_ENABLED("messaging.setDoNotDisturbEnabled"),
    MESSAGING_SHOW_PENDING_MESSAGE("messaging.showPendingMessage"),
    PUSH_GET_LAST_KNOWN_TOKEN("push.getLastKnownPushToken"),
    PUSH_REQUEST_PERMISSION("push.requestPermission"),
    PUSH_IOS_REQUEST_PROVISIONAL_PERMISSION("push.iOS.requestProvisionalPermission"),
    PUSH_IOS_REFRESH_TOKEN("push.iOS.refreshToken"),
    PUSH_IOS_SET_SHOW_FOREGROUND("push.iOS.setShowForegroundNotifications"),
    PUSH_CLEAR_BADGE("push.clearBadge"),
    PUSH_DISMISS_NOTIFICATIONS("push.dismissNotifications"),
	USER_EDIT("user.edit"),
	USER_TRACK_EVENT("user.track.event"),
	USER_TRACK_TRANSACTION("user.track.transaction"),
    USER_TRACK_LOCATION("user.track.location"),
    USER_GET_IDENTIFIER("user.getIdentifier"),
    USER_GET_LANGUAGE("user.getLanguage"),
    USER_GET_REGION("user.getRegion"),
	USER_GET_INSTALLATION_ID("user.getInstallationID"),
    USER_FETCH_ATTRIBUTES("user.fetch.attributes"),
    USER_FETCH_TAGS("user.fetch.tags"),
    DEBUG_SHOW_DEBUG_VIEW("debug.showDebugView"),

    INBOX_CREATE_INSTALLATION_FETCHER("inbox.createInstallationFetcher"),
    INBOX_CREATE_USER_FETCHER("inbox.createUserFetcher"),
    INBOX_RELEASE_FETCHER("inbox.releaseFetcher"),
    INBOX_FETCH_NEW_NOTIFICATIONS("inbox.fetchNewNotifications"),
    INBOX_FETCH_NEXT_PAGE("inbox.fetchNextPage"),
    INBOX_GET_FETCHED_NOTIFICATIONS("inbox.getFetchedNotifications"),
    INBOX_MARK_AS_READ("inbox.markAsRead"),
    INBOX_MARK_ALL_AS_READ("inbox.markAllAsRead"),
    INBOX_MARK_AS_DELETED("inbox.markAsDeleted"),
    INBOX_DISPLAY_LANDING("inbox.displayLandingMessage"),

    /// For testing
    ECHO("echo");

    /**
     * Action name, the one passed as a paremeter in {@link com.batch.batch_flutter.interop.BatchBridge}
     */
    private String name;

    /**
     * Init an action with its string representation
     */
    private Action(String name)
    {
        this.name = name;
    }

    /**
     * Get the name associated with this action
     */
    public String getName()
    {
        return name;
    }

    /**
     * Create an Action from its string name.
     *
     * @throws IllegalArgumentException Thrown when the supplied name doesn't match any known action
     * @param actionName Action name
     * @return Action
     */
    public static Action fromName(String actionName) throws IllegalArgumentException
    {
        for (Action parameterValue : values())
        {
            if (parameterValue.getName().equalsIgnoreCase(actionName))
            {
                return parameterValue;
            }
        }

        throw new IllegalArgumentException("Unknown action.");
    }
}
